lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,307 @@
+    _addTrueCondition(node.condition);
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addFalseCondition(node.condition);
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    final elseScope = _visitElseStatement(node.elseStatement, node.condition);
+    _visitIfStatement(node);
+    _propagateUndefinedExpressions(elseScope);
+    if (_isLastStatementAnExitStatement(node.thenStatement)) {
+      _addFalseCondition(node.condition);
+    }
+    if (_isLastStatementAnExitStatement(node.elseStatement)) {
+      _addTrueCondition(node.condition);
+    }
+  }
+
+  @override
+  visitNode(AstNode node) {
+    node.visitChildren(this);
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    _addElementToEnvironment(new _UndefinedExpression(node.element));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitWhileStatement(WhileStatement node) {
+    _addScope();
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addTrueCondition(node.condition);
+    node.body?.accept(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node.body, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addFalseCondition(node.condition);
+    }
+  }
+
+  void _addElementToEnvironment(_ExpressionBox e) {
+    outerScope?.add(e);
+  }
+
+  void _addFalseCondition(Expression expression) {
+    _addElementToEnvironment(
+        new _ConditionExpression(expression, value: false));
+  }
+
+  void _addScope() {
+    outerScope = new ConditionScope(outerScope);
+  }
+
+  void _addTrueCondition(Expression expression) {
+    _splitConjunctions(expression).forEach((e) {
+      _addElementToEnvironment(new _ConditionExpression(e));
+    });
+  }
+
+  Iterable<Expression> _getExpressions(Iterable<Element> elements,
+      {bool value: true}) {
+    return outerScope.getExpressions(elements, value);
+  }
+
+  bool _isLastStatementAnExitStatement(Statement statement) {
+    if (statement is Block) {
+      return _isLastStatementAnExitStatement(
+          DartTypeUtilities.getLastStatementInBlock(statement));
```

Comment:
A return, a throw, or a rethrow (assuming I'm not forgetting anything).

As long as we've thought hard about possible false positives and worked to prevent them, and document in the code how we've done that, I'm fine with most solutions. (Including undefining everything after an exit from the current function body.)